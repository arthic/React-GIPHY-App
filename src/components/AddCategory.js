import React, {useState} from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({setCategories}) => {
	/* Deja el string vacio para que no te marque error ya que
	si no hay comillas recibe undefined*/
	const [inputValue, setInputValue] = useState('')
	const handleInputChange = (e) => {
		// console.log(e.target.value);
		setInputValue(e.target.value)
		console.log('Cambio el input');
	}
	const handleSubmit = (e) => {
		/*Comportamiento por defecto del formulario es la
		recarga de la pagina*/
		e.preventDefault()
		console.log('handleSubmit', inputValue);

		if (inputValue.trim().length > 2) {
			// Se llama como callback que tiene el estado anterior
			setCategories(cats => [inputValue, ...cats])
			setInputValue('')
		}
	}
	/* const handleFocusChange = (e) => {
		setInputValue(e.target.value)
		// console.log(valorDefault);
		// setInputValue(e.target.value='')
		console.log(e.target.value);
		if (e.target.value === 'Busca un GIF!!!') {
			setInputValue(e.target.value='')
		}
		console.log(e.target.value);
	}*/
	return (
		/* El fragment "<>" agrupa elementos HTML o JSX, al usar
		un Form ya se esta agrupando el elemento, por lo tanto ya
		no es necesario el fragment*/
		<form onSubmit={handleSubmit}>
			<div className="search-contend">
				<i className="fas fa-search"></i>
				<input
					type="text"
					value={inputValue}
					onChange={handleInputChange}
					// onFocus={handleFocusChange}
				/>
			</div>
		</form>
	)
}
AddCategory.propTypes = {
	setCategories: PropTypes.func.isRequired
}
export default AddCategory
