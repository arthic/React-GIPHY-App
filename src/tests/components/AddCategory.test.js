import React  from 'react'
import {shallow} from 'enzyme'
import '@testing-library/jest-dom'
import AddCategory from '../../components/AddCategory';

describe('Pruebas en <AddCategory />', () => {
	// Mejor manejo del testing de la funcion que recibe
	const setCategories = jest.fn()
	let wrapper = shallow(<AddCategory setCategories={setCategories} />)

	// Limpia los test o mooks para evitar conflictos
	beforeEach( () => {
		jest.clearAllMocks()
		wrapper = shallow(<AddCategory setCategories={setCategories} />)
	})
	test('Despliege correcto de <AddCategory />', () => {
		expect(wrapper).toMatchSnapshot()
	})
	test('Debe cambiar caja de texto', () => {
		const input = wrapper.find('input')

		// Definir el evento {e}
		const value = 'Hola Cambio'
		input.simulate('change', {target: {value}})
	})
	test('No manda informaciÃ³n a traves de submit', () => {
		wrapper.find('form').simulate('submit', {
			preventDefault(){}
		})
		// Que no retorne ningun valor en "inputValue"
		expect(setCategories).not.toHaveBeenCalled()
	});
	test('Debe de llamar el setCategories y limpiar la caja de texto', () => {
		const value = 'Hola cambio'
		// 1.- Simular input change
		wrapper.find('input').simulate('change', {target:{value}})
		// 2.- simular el submit
		wrapper.find('form').simulate('submit', { preventDefault(){} })
		// 3.- setCategories se debe haner llamado
		expect(setCategories).toHaveBeenCalled()
		expect(setCategories).toHaveBeenCalledTimes(1)
		expect(setCategories).toHaveBeenCalledWith(expect.any(Function))
		// 4.- El valor del input debe de estar ''
		expect(wrapper.find('input').prop('value')).toBe('')
	})
});
