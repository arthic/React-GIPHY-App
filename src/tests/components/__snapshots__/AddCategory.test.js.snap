// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Despliege correcto de <AddCategory /> 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="search-contend"
  >
    <i
      className="fas fa-search"
    />
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
</form>
`;
